File: src/commands/secret.ts
import { Command } from "https://deno.land/x/cliffy@v0.25.0/command/mod.ts"
import { Secret as SecretPrompt } from "https://deno.land/x/cliffy@v0.25.0/prompt/mod.ts"

import { decrypt, encrypt } from "../aes.ts"
import { Secret } from "../postgres.ts"

const clone = new Command()
	.name("clone")
	.description("Clone the repository secret to the original location")
	.action(async _ => {
		const key = await SecretPrompt.prompt("Enter encryption key: ")
		if (key !== Deno.env.get("AES__KEY")) {
			console.log("Incorrect key")
			return
		}

		const match = Deno.cwd().match(/^C:\\Projects\\([^\\]*)$/)
		if (!match) {
			console.log("Invalid project path")
			return
		}

		const secret = await Secret.where({ project: match[1]! }).first()
		if (!secret) {
			console.log("No secret found for this project")
			return
		}

		Deno.writeTextFileSync(<string>secret.path, await decrypt(<string>secret.content, key))
	})

const set = new Command()
	.name("set")
	.description("Set the repository secret, update if it already exists")
	.arguments("<path>")
	.action(async (_, path) => {
		const key = await SecretPrompt.prompt("Enter encryption key: ")
		if (key !== Deno.env.get("AES__KEY")) {
			console.log("Incorrect key")
			return
		}

		const match = Deno.cwd().match(/^C:\\Projects\\([^\\]*)$/)
		if (!match) {
			console.log("Invalid project path")
			return
		}

		try {
			await Secret.create({
				project: match[1]!,
				path,
				content: await encrypt(Deno.readTextFileSync(path), key)
			})
			console.log("Stored encrypted secret file:", path)
		} catch (err) {
			console.error(err)
		}
	})

const secret = new Command()
	.name("secret")
	.description("Manage secrets for different repositories")
	.action(_ => {
		secret.showHelp()
	})
	.command("clone", clone)
	.command("set", set)

export default secret

File: src/commands/secret.rs
use {
    super::super::{
        aes::{decrypt, encrypt, validate},
        create_connection,
        models::Secret,
        schema::secrets,
    },
    dialoguer::{theme::ColorfulTheme, Password},
    diesel::{insert_into, prelude::*},
    regex::Regex,
    seahorse::Command,
    std::{env::current_dir, fs, path::PathBuf},
};

fn clone() -> Command {
    Command::new("clone")
        .description("Clone the repository secret to the original location")
        .action(|_| {
            let key = Password::with_theme(&ColorfulTheme::default())
                .with_prompt("Enter password: ")
                .interact()
                .unwrap();

            if !validate(&key) {
                println!("Incorrect key");
                return;
            }

            let regex = Regex::new(r#"^C:\\Projects\\([^\\]*)$"#).unwrap();
            if !regex.is_match(current_dir().unwrap().to_str().unwrap()) {
                println!("Invalid project path");
                return;
            }

            let directory = current_dir().unwrap();
            let project = match regex.captures(directory.to_str().unwrap()) {
                Some(captures) => captures.get(1).unwrap().as_str(),
                None => {
                    println!("Invalid project path");
                    return;
                }
            };

            let secret = match secrets::dsl::secrets
                .filter(secrets::project.eq(project))
                .first::<Secret>(&mut create_connection())
            {
                Ok(secret) => secret,
                Err(_) => {
                    println!("No secret found for this project");
                    return;
                }
            };

            match fs::write(
                PathBuf::from(&secret.path),
                decrypt(&secret.content, &key).unwrap(),
            ) {
                Ok(_) => println!("Wrote to file: {}", secret.path),
                Err(err) => {
                    println!("Unable to write file: {}", secret.path);
                    println!("Error: {}", err);
                }
            }
        })
}

fn set() -> Command {
    Command::new("set")
        .description("Set the repository secret, update if it already exists")
        .action(|context| {
            let regex = Regex::new(r#"^C:\\Projects\\([^\\]*)$"#).unwrap();
            if !regex.is_match(current_dir().unwrap().to_str().unwrap()) {
                println!("Invalid project path");
                return;
            }

            let directory = current_dir().unwrap();
            let project = match regex.captures(directory.to_str().unwrap()) {
                Some(captures) => captures.get(1).unwrap().as_str().into(),
                None => {
                    println!("Invalid project path");
                    return;
                }
            };

            let path = match context.args.first() {
                Some(path) => path.to_string(),
                None => {
                    println!("No path provided");
                    return;
                }
            };

            let content = match fs::read_to_string(&path) {
                Ok(content) => content,
                Err(_) => {
                    println!("Unable to read file: {}", path);
                    return;
                }
            };

            let key = Password::with_theme(&ColorfulTheme::default())
                .with_prompt("Enter password: ")
                .interact()
                .unwrap();

            if !validate(&key) {
                println!("Incorrect key");
                return;
            }

            let secret = Secret {
                project,
                path: path.clone(),
                content: encrypt(&content, &key).unwrap(),
            };

            match insert_into(secrets::dsl::secrets)
                .values(&secret)
                .on_conflict(secrets::project)
                .do_update()
                .set(&secret)
                .execute(&mut create_connection())
            {
                Ok(_) => {
                    println!("Secret stored: {}", path);
                }
                Err(err) => {
                    println!("Unable to store secret: {}", path);
                    println!("Error: {}", err);
                }
            }
        })
}

pub fn secret() -> Command {
    Command::new("secret")
        .description("Manage secrets for different repositories")
        .command(clone())
        .command(set())
        .action(|context| context.help())
}
